lib/publiccloud/instance.pm

```diff
@@ -352,40 +352,65 @@ sub wait_for_ssh {
     my $start_time = time();
     my $check_port = 1;
     my $sleep_period = $args{ignore_wrong_pubkey} ? 20 : 1;
-
+    #
+    my @sem = (1, 1, 1);
+    my @out = ('') x 3;
+    my @command = ('sudo systemctl is-system-running --wait', 'systemd-analyze time', 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"');
+    my @expect = ('running|degraded', 'Startup finished in', 'Reached target ');
+    my $expect_ko_0 = 'maintenance|stopping|offline|unknown';    # wrong states of command[0]
+    my $pass = 0;
     # Looping until reaching timeout or passing two conditions :
     # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
+    # - Startup time successfully collected + is-system-running true + journal contains target
+    #   - journalctl got message about reaching one of certain targets
     while ((my $duration = time() - $start_time) < $args{timeout}) {
         if ($check_port) {
             $check_port = 0 if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
         }
         else {
             # On boottime test we do hard reboot which may change the instance address
             script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
+            # here 3 check methods 0 to 2 from @command used:
+            foreach my $j (0 .. 2) {
```

Comment:
I don't know maybe it is something personal problem for me , but I having hard times to "process" this code and understand what is going on here , you have 3 arrays which contains 3 pieces of 3 different flows split by array ID so to understand what is going on in each case you need to take one pieces from each array . I think if we will drop this loop and drop arrays and just write actions one by one it would be much much readable and easier to understand 

```
my $is_sys_run_cmd = 'sudo systemctl is-system-running --wait';
$out_is_sys_run = $self->run_ssh_command(cmd => $is_sys_run_cmd, proceed_on_failure => 1, username => $args{username});
if ($is_sys_run_out =~ m/degraded/) {
                    my $failed = $self->run_ssh_command(cmd => 'sudo systemctl --failed', proceed_on_failure => 1, username => $args{username});
                    record_info("CHECK SSH", 'wait ssh ' . (++$pass) . ' fail.serv.,' . (time() - $start_time) . 's, System:' . "\n" . $failed);
                }
```

Reply:
Please consider to use the [note-5](https://progress.opensuse.org/issues/126218#note-5) for some indication. Moreover I did above an analysis of the last tests results, so it may be tha now I could use one only check type :) .